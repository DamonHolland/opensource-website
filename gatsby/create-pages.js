const path = require(`path`);

const allProjectsQuery = `
  query allProjects {
    allProjects(filter: {projectType: {eq: "newrelic"}}) {
      edges {
        node {
          ...exploreProjectsFields
        }
      }
    }
    allCategories: allProjects {
      group(field: ossCategory___title) {
        fieldValue
        totalCount
      }
    }
    allLanguages: allProjects {
      group(field: stats___languages___name) {
        fieldValue
        totalCount
      }
    }
    
    allProjectTags {
      group(field: title) {
        fieldValue
      }
    }
  }

  fragment exploreProjectsFields on Projects {
    id
    slug
    fullName
    githubUrl
    stats {
      id
      contributors
      commits
      languages {
        id
        name
        color
      }
      releases
      screenshots
    }
    description
    iconUrl
    name
    ossCategory {
      slug
      title
      description
    }
    owner {
      login
      type
    }
    permalink
    primaryLanguage
    projectTags {
      slug
      title
    }
    shortDescription
    supportUrl
    projectTags {
      title
      slug
    }
    title
    version
    website {
      title
      url
    }
  }
`;

const getProjectComponent = projectType => {
  if (projectType === 'external') {
    return path.resolve('./src/templates/external-project-page.js');
  }

  if (projectType === 'external') {
    return path.resolve('./src/templates/partner-project-page.js');
  }

  return path.resolve(`./src/templates/project-page.js`);
};

const createProjectPages = async ({ graphql, actions }) => {
  const { createPage } = actions;
  const result = await graphql(`
    query AllProjects {
      allProjects {
        edges {
          node {
            id
            name
            slug
            fullName
            permalink
            projectType
          }
        }
      }
    }
  `);
  const pages = result.data.allProjects.edges;

  pages.forEach(({ node }) => {
    const component = getProjectComponent(node.projectType);

    // https://www.gatsbyjs.org/docs/actions/#createPage
    createPage({
      path: `projects/${node.fullName}`, // {org/user}/{repository-foo-name}, ex. newrelic/nr1-workload-geoops
      component,
      context: {
        // Data passed to context is available
        // in page queries as GraphQL variables.

        slug: node.slug // {org/user}-{repository-foo-name}, ex. newrelic-nr1-workload-geoops
      }
    });
  });
};

/*
 * Note:
 *    Although our exploreProjectsFields fragment defined in src/fragments/index.js is available in /pages,
 *    it is not automagically available when running graphql here. For now, we've copied it.
 *
 * TO DO:
 *    Investigate if Gatsby has this available somewhere when running the API methods, perhaps
 *    something like graphql.fragments[name of fragment]
 */
const createExploreProjects = async ({ graphql, actions }) => {
  const { createPage } = actions;

  const allProjectsResult = await graphql(allProjectsQuery);
  const {
    allProjects,
    allCategories,
    allLanguages,
    allProjectTags
  } = allProjectsResult.data;

  createPage({
    path: '/explore-projects',
    component: path.resolve(`./src/templates/explore-projects.js`),
    context: {
      projectData: {
        allProjects: allProjects.edges.map(p => p.node),
        allLanguages: allLanguages.group,
        allCategories: allCategories.group,
        allProjectTags: allProjectTags.group,
        // sortBy:
        searchEngineOptions: {
          indexStrategy: 'Prefix match',
          searchSanitizer: 'Lower Case',
          indexFields: [
            ['slug'],
            ['fullName'],
            ['name'],
            ['description'],
            ['ossCategory', 'title'],
            ['website', 'title'],
            ['projectTags', 'title'],
            ['primaryLanguage']

            // Nested array syntax doesn't work until this PR is merged:
            // https://github.com/bvaughn/js-search/pull/78/files#diff-5d56a676b2913ba26d67295fe642d9b0R18
            // ['stats', 'languages', '[]', 'name']
          ],
          removeStopWords: true
        }
      }
    }
  });
};

module.exports = async params => {
  createProjectPages(params);
  createExploreProjects(params);
};
