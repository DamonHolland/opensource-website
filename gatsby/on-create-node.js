// https://www.gatsbyjs.org/docs/actions/#createNode

const { v4: uuidv4 } = require('uuid');

const isMdx = type => type === 'Mdx';
const slugFromAbsoluteFilePath = (dir, fileAbsolutePath) =>
  fileAbsolutePath
    .slice(fileAbsolutePath.indexOf(dir) + dir.length, fileAbsolutePath.length)
    .replace('.mdx', '');

const createProjectMainContent = ({ node, actions }) => {
  const { createNode } = actions;
  const projectMainContentDir = 'src/data/project-main-content/';
  const isMainContent =
    node.fileAbsolutePath.indexOf(projectMainContentDir) >= 0;

  if (!isMainContent) {
    return;
  }

  const fileAbsolutePath = node.fileAbsolutePath;
  const slug = slugFromAbsoluteFilePath(
    projectMainContentDir,
    fileAbsolutePath
  );

  const fieldData = {
    slug
  };
  createNode({
    ...fieldData,

    // Required fields.
    id: uuidv4(),
    parent: node.id,
    children: [],
    internal: {
      type: `ProjectMainContent`,
      contentDigest: node.internal.contentDigest // We can safely reuse this hash since it's content-based
    }
  });
};

module.exports = params => {
  const { node } = params;

  // if (node.internal.type === `projectsJson`) {
  //   const basePath = `projects`;
  //   const slug = createFilePath({ node, getNode, basePath });
  //   createNodeField({
  //     node,
  //     name: `slug`,
  //     value: slug
  //   });
  // }

  // Transform readmes into an object type of ProjectMainContent which we'll stitch into GraphQL in `type-defs.gql`
  // This will enable us to get all data needed for a page with one dynamic query
  if (isMdx(node.internal.type)) {
    createProjectMainContent(params);
  }
};
