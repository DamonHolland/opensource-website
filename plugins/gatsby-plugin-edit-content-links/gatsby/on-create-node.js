// https://www.gatsbyjs.org/docs/actions/#createNode

// TO DO - allow input parameters from gatsby-config.js
const githubUrl = () => {
  const options = {
    host: 'https://github.com',
    org: 'newrelic',
    repo: 'opensource-website',
    defaultBranch: 'develop'
  };

  const { host, org, repo, defaultBranch } = options;
  const githubBaseUrl = [host, org, repo, 'tree', defaultBranch].join('/');
  return githubBaseUrl;
};

const editLinksMap = {
  '/projects/(.*)': `src/data/project-main-content/$match.mdx`,
  '^/$': `src/pages/index.js`
};

const getEditLinkFromPath = ({ path }) => {
  const result = Object.entries(editLinksMap).reduce(
    (previousValue, [expression, value]) => {
      if (previousValue.link) {
        return previousValue;
      }

      const match = path.match(expression);

      if (Array.isArray(match) && match[0]) {
        if (value.indexOf('$match') >= 0) {
          return { link: value, slug: match[1] };
        }
        return { link: value, slug: '' };
      }

      return previousValue;
    },
    { link: false, slug: '' }
  );

  if (result.link) {
    if (result.link.indexOf('$match') >= 0) {
      const link = result.link.replace('$match', result.slug.replace('/', '-'));
      return link;
    }
    const link = result.link;
    return link;
  }
};

module.exports = params => {
  const { node, actions } = params;
  const { createNodeField } = actions;
  const { path, componentPath } = node;
  const cwd = process.cwd();

  if (node.internal.type === 'SitePage') {
    if (!path || !componentPath) {
      // console.log(JSON.stringify(node, null, 2));
      return;
    }
    const githubBaseUrl = githubUrl();
    const mappedEditLink = getEditLinkFromPath({
      path
    });

    const sourceFilePath = componentPath.replace(`${cwd}/`, '');
    const editLink = !mappedEditLink ? sourceFilePath : mappedEditLink;

    const fullUrl = [githubBaseUrl, editLink].join('/');
    // console.log(fullUrl);

    createNodeField({
      node,
      name: `contentEditLink`,
      value: fullUrl
    });
  }
};
