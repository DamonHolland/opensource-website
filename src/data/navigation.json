{
  "navigation": [
    {
      "label": "Open Standards",
      "link": "/open-standards",
      "order": 1
    },
    {
      "label": "Explore Projects",
      "link": "/explore-projects",
      "order": 2
    },
    {
      "label": "Changelog",
      "link": "/blog",
      "order": 3
    }
  ]
}