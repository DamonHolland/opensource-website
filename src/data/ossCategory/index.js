[
  {
    "title": "New Relic One Catalog Project",
    "slug": "new-relic-one-catalog-project",
    "description": "This is code that will be included in the New Relic One Catalog. It must meet certain criteria which include and supercede the requirements for a Community Project to qualify for this category. Among them is that the maintainer team will be expected to maintain a public roadmap for the project."
  },
  {
    "title": "Community Project",
    "slug": "community-project",
    "description": "This is code that is developed in the open with input from the community through issues and PRs. There must be an active maintainer team that complies with internal New Relic policy on maintainer expectations. There will be a support Topic in the New Relic Explorers Hub and documentation should be reviewed by the New Relic documentation team."
  },
  {
    "title": "Example Code",
    "slug": "example-code",
    "description": "This is code that demonstrates how to do something on the New Relic One platform and is attached to a piece of published content (a blog post, video, etc.). It has a complete description of its usage in the README and any other relevant docs. There is no long-term maintainer for this code (though there is an author who can be addressed with future updates / needs as platform features change). There will be no New Relic Explorers Hub Topic for support."
  },
  {
    "title": "Product delivered in Open Source",
    "slug": "product-delivered-in-open-source",
    "description": "This is code that is developed and maintained by New Relic engineering teams (often but not always) in a private code repository but is delivered via public GitHub."
  },
  {
    "title": "New Relic Experimental",
    "slug": "new-relic-experimental",
    "description": "This is code that solves an interesting problem but there is no intent to assign an active maintainer. The code may be developed in the open for the purpose of rapid feedback on a new feature or function."
  },
  {
    "title": "Archived",
    "slug": "archived",
    "description": "The code is read-only. There is no maintainer team nor support."
  }
]