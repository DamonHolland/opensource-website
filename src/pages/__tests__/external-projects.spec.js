import React from 'react';
import TestRenderer from 'react-test-renderer'; // https://reactjs.org/docs/test-renderer.html
// import { render } from '@testing-library/react'; // https://testing-library.com/docs/react-testing-library/intro
import { useStaticQuery } from 'gatsby';

import ExternalProjectsPage from '../external-projects';
import data from './fixtures/external-projects';

beforeEach(() => {
  useStaticQuery.mockImplementationOnce(() => ({
    site: {
      siteMetadata: {
        title: 'New Relic Open Source',
        description:
          "New Relic <3's open source. We built this site to make it easy for you to explore the projects we're maintaining as well as our involvement in open standards.",
        author: '@newrelic'
      }
    }
  }));
});

describe('External Projects Page', () => {
  it('Renders correctly', () => {
    const tree = TestRenderer.create(
      <ExternalProjectsPage data={data} />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
