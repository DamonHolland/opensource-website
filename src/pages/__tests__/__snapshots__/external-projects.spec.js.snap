// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`External Projects Page Renders correctly 1`] = `
<div
  className="layout-container undefined"
>
  <div
    className="globalHeaderContainer  "
  >
    <div
      className="globalHeaderContent"
    >
      <div
        className="leftSideLinksContainer"
      >
        <a
          className="logo"
          href="https://newrelic.com/"
          rel="noopener noreferrer"
        >
          New Relic
        </a>
        <ul
          className="leftSideLinks"
        >
          <li
            className="leftSideLinkItem"
          >
            <a
              className="leftSideLink"
              href="https://developer.newrelic.com/"
            >
              Developers
            </a>
          </li>
          <li
            className="leftSideLinkItem active"
          >
            <a
              className="leftSideLink"
              href="#"
            >
              Open Source
            </a>
          </li>
          <li
            className="leftSideLinkItem"
          >
            <a
              className="leftSideLink"
              href="https://docs.newrelic.com/"
            >
              Docs
            </a>
          </li>
        </ul>
      </div>
      <ul
        className="rightSideButtons"
      >
        <li
          className="darkModeToggle"
        />
        <li
          className="rightSideButton githubButton"
        >
          <a
            className="githubButtonLink"
            href="https://github.com/newrelic/opensource-website"
          >
            GitHub
          </a>
        </li>
      </ul>
    </div>
  </div>
  <header
    className="primaryHeaderContainer  "
  >
    <a
      className="primaryHeaderLogo"
      href="/"
    >
      New Relic Open Source
    </a>
    <ul
      className="primaryHeaderNavLinks"
    >
      <li
        className="primaryHeaderNavLinkItem"
      >
        <a
          className="primaryHeaderNavLink"
          href="/external-projects"
        >
          External Projects
        </a>
      </li>
      <li
        className="primaryHeaderNavLinkItem"
      >
        <a
          className="primaryHeaderNavLink"
          href="/explore-projects"
        >
          New Relic Projects
        </a>
      </li>
    </ul>
    <div
      className="primaryHeaderMobileNav "
    >
      <button
        className="mobileMenuButton"
        onClick={[Function]}
      >
        Menu
      </button>
      <ul
        className="primaryHeaderMobileNavLinks"
      >
        <li
          className="primaryHeaderNavLinkItem"
        >
          <a
            className="primaryHeaderNavLink"
            href="/external-projects"
          >
            External Projects
          </a>
        </li>
        <li
          className="primaryHeaderNavLinkItem"
        >
          <a
            className="primaryHeaderNavLink"
            href="/explore-projects"
          >
            New Relic Projects
          </a>
        </li>
      </ul>
    </div>
  </header>
  <main
    className="layout-container-main full-width undefined"
  >
    <div
      className="featuredProjects"
    />
  </main>
  <footer
    className="footerContainer"
  >
    <div
      className="footerContents"
    >
      <div
        className="footerLogoContainer"
      >
        <a
          className="footerLogo hasEditButton"
          href="/"
        >
          New Relic Open Source
        </a>
        <a
          className="editButton"
          href="https://github.com/newrelic/opensource-website/tree/develop/src/pages/code-of-conduct.mdx"
          rel="noopener noreferrer"
          target="__blank"
        >
          <img
            alt="pencil icon"
            className="editButtonIcon"
            src="test-file-stub"
          />
          Edit this page
        </a>
      </div>
      <div
        className="meta"
      >
        <small
          className="version"
        >
          Version 
          0.5.0
        </small>
        <small
          className="footerLegal"
        >
          Copyright Â© 
          2020
           New Relic Inc.
        </small>
      </div>
      <ul
        className="footerNavLinks"
      >
        <li
          className="footerNavLinkItem"
        >
          <a
            className="footerNavLink"
            href="/standards"
          >
            Standards
          </a>
        </li>
        <li
          className="footerNavLinkItem"
        >
          <a
            className="footerNavLink"
            href="/external-projects"
          >
            External Projects
          </a>
        </li>
        <li
          className="footerNavLinkItem"
        >
          <a
            className="footerNavLink"
            href="/explore-projects"
          >
            New Relic Projects
          </a>
        </li>
      </ul>
    </div>
  </footer>
</div>
`;
